### DISCUSSION TITLE
How to authenticate for AddContainer

### DISCUSSION THREAD

**Discussion by JohnVottero** [microsoft/aspire#3440] [Q&A]
I'm trying to use AddContainer(...) to add a container service in my Aspire AppHost. The container is in an Azure Container Registry. When I run the AppHost it fails with:

```
fail: Aspire.Hosting.Dcp.dcpctrl.ContainerReconciler[0]
      could not create the container    {"Container": {"name":"tenantServicesService"}, "Reconciliation": 2, "error": "CreateContainer command returned error code 1 Stdout: '' Stderr: 'Unable to find image 'crtenantdev.azurecr.io/tenant-services-service:latest' locally\nError response from daemon: Head \"https://crtenantdev.azurecr.io/v2/tenant-services-service/manifests/latest\": unauthorized: {\"errors\":[{\"code\":\"UNAUTHORIZED\",\"message\":\"authentication required, visit https://aka.ms/acr/authorization for more information.\"}]}\n'"}
```
I tried adding a call to AddAzureProvisioning() in my AppHost thinking that would pickup the credentials from Visual Studio's Azure Service Authentication (where I am authenticated) but that did not help.

How can I add credentials for a container registry? 

This is with Aspire Preview 4.

**Comment by davidfowl** [comment]
AFAIK we don't have support for pulling from private registries in development. You may be able to work around this by manually pulling this image in your local registry (and entering credentials there to do so), then re-running the application.

cc @karolz-ms

**Reply by JohnVottero** [reply]
This workaround looks promising. I pulled the containers by doing:

```
az login
az acr login --name my-acr-repo-name
docker pull my-acr-repo-name/my-image-name
```
When I start my AppHost the containers start!

They're currently failing because they don't have TLS certificates but, that's a completely different problem.

**Comment by marta-beznos** [comment]
Is the support for this functionality planned? It would be convenient if Aspire could auto-pull the image from the remote registry.

**Reply by davidfowl** [reply]
No it's not, can you file an issue?

**Comment by mortenn** [comment]
My workaround was to use CliWrap in the AppHost;
```C#
await Cli.Wrap("az")
  .WithArguments($"acr login --name {registry}")
  .WithStandardOutputPipe(PipeTarget.ToDelegate(Console.Out.WriteLine))
  .WithStandardErrorPipe(PipeTarget.ToDelegate(Console.Error.WriteLine))
  .WithValidation(CommandResultValidation.None)
  .ExecuteAsync();
```
